### PR TITLE
CI Build&Test: Replace use of `sage -t --new`

### LINKED ISSUE #37985: CI Build&Test: Replace use of `sage -t --new` by explicit file list
The current method, using `git reset` to mark what's new in the PR and then relying on `sage -t --new`, does not handle file deletions/renames correctly, as observed in https://github.com/sagemath/sage/pull/37839#issuecomment-2103562852

Instead we should obtain the list of files changed relative to the base of the PR.

### PR DESCRIPTION
**Pull request by mkoeppe**
Fixes #37985.

We also turn on  `--long --format github` for this run, so that annotations are shown in the "Files changed" tab.
(Previously, only the annotations from the full testsuite run were shown, out of the concern of duplicating the numerous doctest style warnings. But these warnings have been much reduced, and the last two warnings are fixed in #38029.)

Example run: https://github.com/sagemath/sage/actions/runs/9198070834/job/25299950120?pr=37988

### :memo: Checklist

- [x] The title is concise and informative.
- [ ] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation and checked the documentation preview.

### :hourglass: Dependencies

- Depends on #37926 (merged here)

### CODE REVIEW

File: .ci/write-dockerfile.sh [hunk 1]
```diff
@@ -272,13 +272,14 @@ $ADD src/Pipfile.m4 src/pyproject.toml src/requirements.txt.m4 src/setup.cfg.m4
 $ADD m4 /new/m4
 $ADD pkgs /new/pkgs
 $ADD build /new/build
-$ADD .ci /new/.ci
 $ADD .upstream.d /new/.upstream.d
+ADD .ci /.ci
 RUN if [ -d /sage ]; then                                               \
```
Comment (kwankyu) [inline .ci/write-dockerfile.sh:277]: What's the difference between `$ADD` and `ADD`?

Comment (kwankyu) [reply .ci/write-dockerfile.sh:277]: Ah it is `ADD="ADD $__CHOWN"`...

Comment (kwankyu) [reply .ci/write-dockerfile.sh:277]: Why here `ADD` and `$ADD` below (line 341)?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -175,11 +175,20 @@ jobs:
...
+          files_ignore: src/{setup,conftest*}.py
+
       - name: Test changed files (sage -t --new)
+        if: steps.changed-files.outputs.all_changed_files
         run: |
           export MAKE="make -j2 --output-sync=recurse" SAGE_NUM_THREADS=4
           # We run tests with "sage -t --new"; this only tests the uncommitted changes.
-          ./sage -t --new -p4
+          # https://github.com/tj-actions/changed-files?tab=readme-ov-file#outputs-
+          ./sage -t --long --format github -p4 ${{ steps.changed-files.outputs.all_changed_files }}
```
Comment (kwankyu) [inline .github/workflows/build.yml:191]: Modify line 189 as well?

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -175,11 +175,20 @@ jobs:
           ./sage -python -m pytest -c tox.ini -qq --doctest --collect-only || true
         shell: sh .ci/docker-exec-script.sh BUILD /sage {0}
 
+      - name: Get changed files
+        id: changed-files
+        uses: tj-actions/changed-files@v44
+        with:
+          files: src/**/*.{py,pyx,pxd,pxi,sage,spyx,rst,tex}
+          files_ignore: src/{setup,conftest*}.py
```
Comment (kwankyu) [inline .github/workflows/build.yml:184]: Is `tex` file also tested?

File: .ci/write-dockerfile.sh (hunk 1 above)
Comment (mkoeppe) [reply .ci/write-dockerfile.sh:277]: Thanks for catching this; fixed in f4f0a0ca80f

File: .github/workflows/build.yml (hunk 3 above)
Comment (mkoeppe) [reply .github/workflows/build.yml:184]: I've added a comment

Comment (kwankyu) [reply .github/workflows/build.yml:184]: OK. It is status quo...

**Review by kwankyu — APPROVED** [review]
OK. LGTM.

**Comment by mkoeppe** [comment]
Thanks!

**Pull request merged**